### ISSUE TITLE
E_FAIL while crossgen a method with awaitable

### ISSUE THREAD

**Issue by echesakovMSFT** [dotnet/coreclr#20066]
I am seeing many errors during crossgen of libraries in Core_Root that all looks like these
```
Error 2147500037 (Exception from HRESULT: 0x80004005 (E_FAIL)) while compiling method Microsoft.CodeAnalysis.Diagnostics.AnalyzerDriver.AttachQueueAndProcessAllEventsAsync
Error 2147500037 (Exception from HRESULT: 0x80004005 (E_FAIL)) while compiling method Microsoft.CodeAnalysis.Diagnostics.AnalyzerDriver.ExecutePrimaryAnalysisTaskAsync
Error 2147500037 (Exception from HRESULT: 0x80004005 (E_FAIL)) while compiling method Microsoft.CodeAnalysis.Diagnostics.AnalyzerDriver.ProcessCompilationEventsAsync
Error 2147500037 (Exception from HRESULT: 0x80004005 (E_FAIL)) while compiling method Microsoft.CodeAnalysis.Diagnostics.AnalysisState.OnCompilationEventsGeneratedAsync
Error 2147500037 (Exception from HRESULT: 0x80004005 (E_FAIL)) while compiling method Microsoft.CodeAnalysis.Diagnostics.AnalysisState.EnsureAnalyzerActionCountsInitializedAsync
Error 2147500037 (Exception from HRESULT: 0x80004005 (E_FAIL)) while compiling method Microsoft.CodeAnalysis.Diagnostics.CompilationWithAnalyzers.ComputeAnalyzerDiagnosticsAsync
Error 2147500037 (Exception from HRESULT: 0x80004005 (E_FAIL)) while compiling method Microsoft.CodeAnalysis.Diagnostics.CompilationWithAnalyzers.ComputeAnalyzerDiagnosticsCoreAsync
Error 2147500037 (Exception from HRESULT: 0x80004005 (E_FAIL)) while compiling method Microsoft.CodeAnalysis.Diagnostics.CompilationWithAnalyzers.WaitForActiveAnalysisTasksAsync
Error 2147500037 (Exception from HRESULT: 0x80004005 (E_FAIL)) while compiling method Microsoft.CodeAnalysis.Diagnostics.CompilationWithAnalyzers.PopulateEventsCacheAsync
Error 2147500037 (Exception from HRESULT: 0x80004005 (E_FAIL)) while compiling method <>c__DisplayClass51_2.<ComputeAnalyzerDiagnosticsAsync>b__1
```
Crossgen ignores them and at the end says that native image generated successfully.

It seems that all failure occur at the exactly same place in coreclr https://github.com/dotnet/coreclr/blob/4b9a9d71b06fbce71c7d000e8b2ed3623976b30c/src/vm/ceeload.cpp#L5870

I could reproduce the issue on Windows and Linux on x64 and arm (and maybe others) and the smallest repro I could get is the following
```c#
using System;
using System.Threading.Tasks;

namespace CrossGenFailure
{
    class Program
    {
        static async Task DelayAsync()
        {
            await Task.Delay(1000);
        }

        static void Main(string[] args)
        {
            Task.Delay(1000).Wait();
            Console.WriteLine("Hello World!");
        }
    }
}
```

If I run
```
%Core_Root%\crossgen.exe /Platform_Assemblies_Paths %Core_Root% /out CrossGenFailure.ni.dll /in D:\git\CrossGenFailure\bin\Release\netcoreapp2.1\CrossGenFailure.dll
```
I would get 
```
(Exception from HRESULT: 0x80004005 (E_FAIL)) while compiling method CrossGenFailure.Program.DelayAsync
```
The call stack during the failure 
```
>	crossgen.exe!Module::LookupTypeRefByMethodTable(MethodTable * pMT) Line 5870	C++
 	crossgen.exe!ZapSig::GetSignatureForTypeHandle(TypeHandle handle, SigBuilder * pSigBuilder) Line 209	C++
 	crossgen.exe!CEECompileInfo::EncodeClass(CORINFO_MODULE_STRUCT_ * referencingModule, CORINFO_CLASS_STRUCT_ * classHandle, SigBuilder * pSigBuilder, void * pEncodeModuleContext, unsigned long(*)(void *, CORINFO_MODULE_STRUCT_ *) pfnEncodeModule) Line 1592	C++
 	crossgen.exe!ZapImportTable::GetCheckTypeLayoutImport(CORINFO_CLASS_STRUCT_ * handle) Line 1578	C++
 	crossgen.exe!ZapInfo::getClassSize(CORINFO_CLASS_STRUCT_ * cls) Line 3299	C++
 	clrjit.dll!Compiler::lvaCanPromoteStructType(CORINFO_CLASS_STRUCT_ * typeHnd, Compiler::lvaStructPromotionInfo * StructPromotionInfo, bool sortFields) Line 1667	C++
 	clrjit.dll!Compiler::lvaCanPromoteStructVar(unsigned int lclNum, Compiler::lvaStructPromotionInfo * StructPromotionInfo) Line 1835	C++
 	clrjit.dll!Compiler::fgPromoteStructs() Line 17071	C++
 	clrjit.dll!Compiler::fgMorph() Line 16914	C++
 	clrjit.dll!Compiler::compCompile(void * * methodCodePtr, unsigned long * methodCodeSize, JitFlags * compileFlags) Line 4599	C++
 	clrjit.dll!Compiler::compCompileHelper(CORINFO_MODULE_STRUCT_ * classPtr, ICorJitInfo * compHnd, CORINFO_METHOD_INFO * methodInfo, void * * methodCodePtr, unsigned long * methodCodeSize, JitFlags * compileFlags, CorInfoInstantiationVerification instVerInfo) Line 6085	C++
 	[Inline Frame] clrjit.dll!Compiler::compCompile::__l47::__Body::Run(Compiler::compCompile::__l2::__JITParam *) Line 5440	C++
 	clrjit.dll!Compiler::compCompile(CORINFO_METHOD_STRUCT_ * methodHnd, CORINFO_MODULE_STRUCT_ * classPtr, ICorJitInfo * compHnd, CORINFO_METHOD_INFO * methodInfo, void * * methodCodePtr, unsigned long * methodCodeSize, JitFlags * compileFlags) Line 5444	C++
 	[Inline Frame] clrjit.dll!jitNativeCode::__l8::__Body::Run::__l6::__Body::Run(jitNativeCode::__l8::__Body::Run::__l5::__JITParam *) Line 6722	C++
 	clrjit.dll!`jitNativeCode'::`8'::__Body::Run(jitNativeCode::__l2::__JITParam * __JITpParam) Line 6726	C++
 	clrjit.dll!jitNativeCode(CORINFO_METHOD_STRUCT_ * methodHnd, CORINFO_MODULE_STRUCT_ * classPtr, ICorJitInfo * compHnd, CORINFO_METHOD_INFO * methodInfo, void * * methodCodePtr, unsigned long * methodCodeSize, JitFlags * compileFlags, void * inlineInfoPtr) Line 6750	C++
 	clrjit.dll!CILJit::compileMethod(ICorJitInfo * compHnd, CORINFO_METHOD_INFO * methodInfo, unsigned int flags, unsigned char * * entryAddress, unsigned long * nativeSizeOfCode) Line 310	C++
 	crossgen.exe!ZapInfo::CompileMethod() Line 489	C++
 	crossgen.exe!ZapImage::TryCompileMethodWorker(CORINFO_METHOD_STRUCT_ * handle, unsigned int md, unsigned int methodProfilingDataFlags) Line 2147	C++
 	crossgen.exe!ZapImage::TryCompileMethodDef(unsigned int md, unsigned int methodProfilingDataFlags) Line 1961	C++
 	crossgen.exe!ZapImage::CompileColdRegion() Line 1741	C++
 	crossgen.exe!ZapImage::Compile() Line 1801	C++
 	crossgen.exe!Zapper::CompileModule(CORINFO_MODULE_STRUCT_ * hModule, IMetaDataAssemblyEmit * pAssemblyEmit) Line 1585	C++
 	crossgen.exe!Zapper::CompileAssembly(_GUID * pNativeImageSig) Line 1471	C++
 	crossgen.exe!Zapper::CompileInCurrentDomain(const wchar_t * string, _GUID * pNativeImageSig) Line 1213	C++
 	[Inline Frame] crossgen.exe!Zapper::InvokeDomainCallback(Zapper::DomainCallback *) Line 860	C++
 	crossgen.exe!Zapper::Compile(const wchar_t * string, _GUID * pNativeImageSig) Line 1143	C++
 	crossgen.exe!NGenWorker(const wchar_t * pwzFilename, unsigned long dwFlags, const wchar_t * pwzPlatformAssembliesPaths, const wchar_t * pwzTrustedPlatformAssemblies, const wchar_t * pwzPlatformResourceRoots, const wchar_t * pwzAppPaths, const wchar_t * pwzOutputFilename, const wchar_t * pwzPlatformWinmdPaths, ICorSvcLogger * pLogger, const wchar_t * pwszCLRJITPath) Line 120	C++
 	crossgen.exe!wmain(int argc, wchar_t * * argv) Line 924	C++
```
where `pMT` has the following value
```
-		pMT	0x0000024914f8ae88 {m_dwFlags=17039380 m_BaseSize=24 m_wFlags2=1033 ...}	MethodTable *
		m_dwFlags	17039380	unsigned long
		m_BaseSize	24	unsigned long
		m_wFlags2	1033	unsigned short
		m_wToken	1803	unsigned short
		m_wNumVirtuals	4	unsigned short
		m_wNumInterfaces	0	unsigned short
+		debug_m_szClassName	0x0000024914f7e788 "System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1[VoidTaskResult]"	const char *
+		m_pParentMethodTable	{m_addr=2512881796944 }	IndirectPointer<MethodTable *>
+		m_pLoaderModule	{m_delta={m_val=18446744073681678712 } }	RelativePointer<Module *>
+		m_pWriteableData	{m_ptr=2512907710376 }	PlainPointer<MethodTableWriteableData *>
+		m_pEEClass	{m_ptr=2512907658832 }	PlainPointer<EEClass *>
+		m_pCanonMT	{m_ptr=2512907658832 }	PlainPointer<unsigned __int64>
+		m_pPerInstInfo	{m_ptr=2512907710192 }	PlainPointer<PlainPointer<Dictionary *> *>
		m_ElementTypeHnd	2512907710192	unsigned __int64
		m_pMultipurposeSlot1	2512907710192	unsigned __int64
+		m_pInterfaceMap	{m_ptr=56 }	PlainPointer<InterfaceInfo_t *>
		m_pMultipurposeSlot2	56	unsigned __int64
```

**My question:** is this legit failure and we expect VM throw exception at this place? Or not, and the issue needs more deep investigation?

**Comment by jkotas** [comment]
Yes, this looks like a bug
cc @fadimounir

**Comment by fadimounir** [comment]
Yes looks like a bug in the encoding logic. I'll work on a fix (probably after 2~3 weeks since i'll be on a short vacation)

**Comment by Eilon** [comment]
Looks like we're seeing this on ASP.NET Core builds as well on Linux MUSL (but possibly elsewhere as well). Causing thousands of errors. It seems our build currently ignores these errors, so this doesn't appear to be blocking at this time.

**Comment by fadimounir** [comment]
Fixed with #24386. Not exactly a bug, but cross-module encoding of typerefs is not yet supported (future work). We should have been throwing a E_NOTIMPL instead of E_FAIL.
